snapcraft/plugins/qmake.py

```diff
@@ -0,0 +1,139 @@
+    - qt-version:
+      (enum, 'qt4' or 'qt5')
+      Version of Qt to use with qmake.
+    - project-files:
+      (list of strings)
+      list of .pro files to pass to the qmake invocation.
+"""
+
+import os
+import shutil
+
+import snapcraft
+from snapcraft import common
+
+
+class QmakePlugin(snapcraft.BasePlugin):
+
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema['properties']['options'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+        schema['properties']['qt-version'] = {
+            'enum': ['qt4', 'qt5'],
+        }
+        schema['properties']['project-files'] = {
+            'type': 'array',
+            'minitems': 1,
+            'uniqueItems': True,
+            'items': {
+                'type': 'string',
+            },
+            'default': [],
+        }
+
+        # Qt version must be specified
+        schema['required'].append('qt-version')
+
+        # Inform Snapcraft of the properties associated with building and
+        # pulling so it can mark those steps dirty if they change in the YAML.
+        schema['build-properties'].append('options')
+        schema['pull-properties'].append('qt-version')
+
+        return schema
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self.build_packages.append('make')
+        if self.options.qt_version == 'qt5':
+            self.build_packages.extend(['qt5-qmake', 'qtbase5-dev'])
+        elif self.options.qt_version == 'qt4':
+            self.build_packages.extend(['qt4-qmake', 'libqt4-dev'])
+        else:
+            raise RuntimeError('Unsupported Qt version: {!r}'.format(
+                self.options.qt_version))
+
+    def build(self):
+        if os.path.exists(self.builddir):
+            shutil.rmtree(self.builddir)
+        os.mkdir(self.builddir)
+
+        source_subdir = getattr(self.options, 'source_subdir', None)
+        if source_subdir:
+            sourcedir = os.path.join(self.sourcedir, source_subdir)
+        else:
+            sourcedir = self.sourcedir
+
+        env = self._build_environment()
+
+        sources = [sourcedir]
+        if self.options.project_files:
+            sources = [os.path.join(sourcedir, project_file)
+                       for project_file in self.options.project_files]
+
+        self.run(['qmake'] + self._extra_config() + self.options.options +
+                 sources, env=env)
+
+        self.run(['make', '-j{}'.format(
+            self.project.parallel_build_count)], env=env)
+
+        self.run(['make', 'install', 'INSTALL_ROOT=' + self.installdir],
+                 env=env)
+
+    def _extra_config(self):
+        extra_config = []
+
+        for root in [self.installdir, self.project.stage_dir]:
```

Comment:
should be mostly fine as this is what we do in the machinery already. Just need to make sure `installdir` is preferred over `stagedir`.